### ISSUE TITLE
C# generated code not suitable for Windows Apps

### ISSUE THREAD

**Issue by Gititg** [swig/swig#455]
Hi,

I am trying to write a C#\XAML universal app, that uses some shared C++ code. I tried wrapping it with SWIG, but even though the generated code compiles well in a regular desktop app, it does not compile in a Windows app. You can try it with the C# example solution, just call the C# code from a PCL or a C#/XAML app. 
The errors are mainly around missing classes, mainly: global::System.Runtime.InteropServices.HandleRef (apparently it is not available in Windows App / WinRT API), missing exceptions (System.ApplicationException & System.SystemException) and some of the code used for director feature (some reflection APIs are missing).
I currently worked around it by temporarily defining all missing APIs myself (e.g.: created HandleRef by inheriting from SafeHandle), but it required overriding with the original namespace and some pretty ugly code. 
Is there a better way to do it? 
Is "C# for WinRT/Windows apps" scenario supported buy SWIG (I have version 3.0.5)? Do you plan to add such support or at least a SWIG flag for it?

Thanks,
Gitit

**Comment by wsfulton** [comment]
See #423. 

SWIG is an all volunteer community project so if you want a new feature, chances are you'll have to adapt SWIG yourself. Probably the thing to do is use SafeHandle instead of HandleRef if this is supported in Windows Apps and this should be doable by overriding a few typemaps. Patches should be submitted to Github.

**Comment by jas99** [comment]
@Gititg Hey thr
Can you share you preety ugly hack :)

**Comment by jas99** [comment]
@wsfulton Hi Mr. Fulton
Do you see any hope for fix for this in near horizon?

Thx

**Comment by wsfulton** [comment]
I'm afraid I don't have the time to work on this. As I said earlier, SWIG is an open source project with multiple contributors and if you really need a new feature you'll need to do it yourself and submit the patch for inclusion in future releases. Or offer to fund the development of it.

**Comment by jas99** [comment]
@wsfulton :) Anyways thx for reply.

**Comment by jas99** [comment]
@Gititg 
@wsfulton Hi Mr. Fulton
I have spent some time looking into the above issue and willing to contribute patch for it. I tried using GCHandle instead of HandleRef. But when I directly store and pass IntPtr in extern function defs it still works. I have read issues with garbage collection but I didn't got any issues in my experiments. What do you think of using IntPtr directly?

Thx

**Comment by paleozogt** [comment]
btw it seems that ApplicationException is deprecated: http://stackoverflow.com/a/5685943/89218

**Comment by wsfulton** [comment]
The first implementation of SWIG used IntPtr and this was rather quickly shown to be buggy as the garbage collector in did sometimes collect the proxy class owning the underlying C++ pointer while a long running method on the proxy class was being invoked. Certainly I observed this in .NET 1.1 and 2.0. HandleRef was designed to solve this very problem, so I don't see how we can go back to IntPtr. Have a look at http://www.mono-project.com/docs/advanced/pinvoke/#gc-safe-pinvoke-code. If the garbage collector has changed behaviour recently, then I'd want to see some documentation on this from Microsoft. From what I recall HandleRef was meant to be buggy and SafeHandle was invented as another solution, but in practice, I have never seen HandleRef not work. Does anyone know what Microsoft's latest solution to the problem is?

**Comment by paleozogt** [comment]
I'm also a bit baffled as to why its been removed in the WinRT version of DotNet, since the `HandleRef` implementation seems totally reasonable.  Unfortunately, there doesn't seem to be a way around it in WinRT.  :(

Perhaps the C# module could take a parameter specifying whether to use `HandleRef` vs `IntPtr`?  It could default to `HandleRef`.

**Comment by wsfulton** [comment]
The generated code is largely from typemaps, so this wouldn't be easy to implement. However, as all the code is from typemaps, you could rewrite the typemaps to use whatever you want. There is just one place that is hard coded to use HandleRef, but that is for director code, so is not relevant unless you are using directors.

I think the best thing is to research what Microsoft's latest suggestion is for a HandleRef replacement in WinRT.

Incidentally, we plan to release swig-3.1 this year and given this will be a major new version, it will be an opportunity to make incompatible changes changing HandleRef to something else if there is a good case for it.

**Comment by ndcomplete** [comment]
Just to add to this discussion since it's been dead for a while, the reference above notes a case where the use of `HandleRef` caused a bug. Basically, when calling P/invoke and passing in a `HandleRef`, the GC reference counter gets decremented before the end of the P/invoke, possibly allowing the `HandleRef` (or the object storing it) to be GC'd. In our case, this caused an Access Violation. Based on my research, it seems that `SafeHandle` fixes this by ensuring that the handle reference is kept alive through the end of the current scope as opposed to allowing it to be greedily GC'd. This fix would also make it safe for use with WinRT, and may also give @paleozogt some solace as to why `HandleRef` was removed.

**Comment by wsfulton** [comment]
Which reference details where SafeHandle fixes a problem that HandleRef doesn't? I'd like to see some code as HandleRef is documented as `Wrapping a handle with HandleRef guarantees that the managed object is not garbage collected until the platform invoke call completes.` at [Microsoft's documentation on HandleRef](https://msdn.microsoft.com/en-us/library/system.runtime.interopservices.handleref(v=vs.110).aspx).

**Comment by ndcomplete** [comment]
@wsfulton Hmmm, the documentation does say that. And based on disassembly, it seems that the library we are using does directly pass the `HandleRef` into the P/invoke method. I'm not sure why it was not working in our case, as we had a directly reproducible case where the reference was being disposed of during a P/invoke call on a system under heavy load. Additionally, adding explicit `GC.KeepAlive` calls after the P/invoke wrappers solved the problem.

Though that said, there are a number of objects that we were using, including a wrapper "parent" object, which we also explicitly kept alive. My current hypothesis is this:

Consider the case where you have a collection object exposed through SWIG: `Collection`
Additionally, you have an 'item' object owned by the collection: `Item`

If you were to do this:

```
Collection collection = new Collection(1, 2, 3);
Item item = collection.Get(1);
// c is now available for collection, as there are no longer any P/invokes nor references to it
item.GetValue(); // Can possibly throw AccessViolationException if collection has been collected
                 // based on how it cleans up its items.
```

My hypothesis of our fix is that, by being over-zealous in our 'KeepAliving' and doing:
```
GC.KeepAlive(collection);
GC.KeepAlive(item);
```

We avoided this issue entirely by happenstance, even if the original thought was that keeping just `item` alive was sufficient.

I thought that I read that `SafeHandle` would solve this as well by ensuring that `collection` does not get garbage collected until it goes out of scope, but I can't find this in the documentation. It seems to work much the same way as `HandleRef` in this regard.

Possible solutions:

- Provide guidance to collections like this to use `GC.KeepAlive`.
- Have the `Item` objects store a reference to the parent `Collection`s to ensure that the parent is kept alive until the child is disposed (and the reference cleared).
- Utilize the already-implemented `IDisposable` pattern on the parent to deterministically dispose of the parent.

[Earlier questions](https://stackoverflow.com/questions/44055101/do-i-need-to-dispose-of-swig-generated-objects) I had asked about the need for the `IDisposable` pattern suggested that it was not necessary. I'll amend these to note the potential value.

**Comment by wsfulton** [comment]
Yes, I would agree that your code could have been subject with the early garbage collection. This is documented in a similar manner at http://www.swig.org/Doc3.0/CSharp.html#CSharp_memory_management_member_variables and the section just after it. Your idea of using a C# _**using statement**_ to make use of the generatedIDisposble is a good one to prevent early garbage collection. I'd appreciate a patch to the documentation in the Doc/Manual/CSharp.html file to add in this idea.

**Comment by wsfulton** [comment]
HandleRef is supported now .NET Core 2.0, see https://github.com/dotnet/coreclr/issues/6758. This was released in 2017. It would be useful if someone could test that it actually works and report back so that this issue can be closed.

**Comment by andyb1979** [comment]
I was able to resolve ApplicationException by changing 

`    static void SetPendingApplicationException(string message) {
      SWIGPendingException.Set(new global::System.ApplicationException(message, SWIGPendingException.Retrieve()));
    }`

to this

`    static void SetPendingApplicationException(string message) {
      SWIGPendingException.Set(new global::System.Exception(message, SWIGPendingException.Retrieve()));
    }`

in csharp/CsharpHead.swg

The HandleRef is defined in csharp.swg. Still working on that one ...

**Comment by andyb1979** [comment]
HandleRef (which I need to be replaced to IntPtr for UWP) can also be resolved by using this in csharp.swg

Before:

`
%typemap(imtype, out="global::System.IntPtr") SWIGTYPE "global::System.Runtime.InteropServices.HandleRef"`

After:
`%typemap(imtype, out="global::System.IntPtr") SWIGTYPE "global::System.IntPtr"`

etc... working on a complete solution now ...
